contracts/delegation/dao-vote-delegation/src/hooks.rs

```diff
@@ -69,97 +64,153 @@ pub(crate) fn execute_nft_stake_changed(
 
-        for LoadedItem {
-            item: Delegation { delegate, percent },
-            expiration,
-            ..
-        } in delegations
-        {
-            // remove the latest delegated VP from the delegate's total and
-            // replace it with the new delegated VP
-
-            // remove original delegated VP if nonzero
-            let current_delegated_vp = calculate_delegated_vp(old_vp, percent);
-            if !current_delegated_vp.is_zero() {
-                remove_delegated_vp(
+    let config = CONFIG.load(deps.storage)?;
+    let old_vp = dao_voting::voting::get_voting_power(
+        deps.as_ref(),
+        delegator,
+        &dao,
+        Some(env.block.height),
+    )?;
+
+    for LoadedItem {
+        item: Delegation { delegate, percent },
+        expiration,
+        ..
+    } in delegations
+    {
+        // for each delegation, we first find the current delegated VP and
+        // the new delegated VP
+        let current_delegated_vp = calculate_delegated_vp(old_vp, percent);
+        let new_delegated_vp = calculate_delegated_vp(new_vp, percent);
+
+        // first we update the next block's delegated VP for the delegate
+        match new_delegated_vp.cmp(&current_delegated_vp) {
+            Ordering::Less => {
+                // if the new delegated VP is lesser than the current delegated VP,
+                // we decrement the delegated VP by the delta.
+                DELEGATED_VP.decrement(
                     deps.storage,
-                    env,
-                    &delegate,
+                    delegate.clone(),
+                    env.block.height + 1,
+                    current_delegated_vp - new_delegated_vp,
+                )?;
+            }
+            Ordering::Equal => {
+                // for cases where current delegated VP is equal to new delegated VP,
+                // we don't need to do anything.
+            }
+            Ordering::Greater => {
+                // if the new delegated VP is greater than the current delegated VP,
+                // we increment the delegated VP by the delta.
+                DELEGATED_VP.increment(
+                    deps.storage,
+                    delegate.clone(),
+                    env.block.height + 1,
+                    new_delegated_vp - current_delegated_vp,
+                )?;
+            }
+        }
+
+        // if original delegation had voting power and an expiration date,
+        // we undo the previous decrement at the end of the expiration period.
+        if current_delegated_vp.u128() > 0 {
+            if let Some(expire_in) = expiration {
+                DELEGATED_VP.increment(
+                    deps.storage,
+                    delegate.clone(),
+                    expire_in,
                     current_delegated_vp,
-                    expiration,
                 )?;
             }
+        }
```

Comment:
i really like the direction of the all changes made to this hook, only incrementing/decrementing the delta instead of fully removing and re-adding. however i _think_ there's a bug here. i'm going to keep looking at it, but with Wormhole, increments and decrements change at the height specified _and_ all future changes, so i'm pretty confident the `increment` here that is undoing the previous expiration needs to take into account the delta as well

Reply:
ok yeah, there's an underflow risk. i fixed it by changing the order from:

- update delta
- undo original expiration
- set new expiration

to:
- undo original expiration
- update delta
- set new expiration

this effectively preserves the order of operations from the original implementation with full remove and then full add, and then combines the updates in the middle into a single delta update

Reply:
yeah great call. even beyond the underflow risk the new order you propose (undo original expiration -> update delta -> set new expiration) is a lot cleaner and easier to follow.